### ISSUE TITLE
Import mechanism in DaphneDSL

### ISSUE THREAD

**Issue by daphne-eu** [daphne-eu/daphne#214]
In GitLab by @pdamme on Mar 8, 2022, 18:20

Like many other script/programming languages, DaphneDSL should have some kind of import/include mechanism enabling users to employ code in other files. This is necessary for both, user files and DaphneDSL standard lib files.

Open questions include how exactly the file to include shall be specified and how exactly it will be included.

**Comment by pdamme** [comment]
This issue has been closed a while ago with #327.

**Comment by pdamme** [comment]
This issue has been closed a while ago with #327.

**Comment by pdamme** [comment]
We need this feature to be able to reuse DaphneDSL code in other DaphneDSL scripts. For instance, we want to create a DAPHNE standard library of high-level data analysis and ML functions, which are in turn implemented as functions in DaphneDSL. Then, users should be able to simply import those library functions and use them in their own DaphneDSL scripts.

It could look like this:
```
import "kmeans.daphne";
in = readMatrix("someFile.csv");
print(kmeans(in, 5, 20));
```

The particulars of this feature are yet to be designed. Here is some inspiration:
- Importing a DaphneDSL script file should have roughly the same effect as pasting its contents literally.
- An exception to this could be that function and variable names could be prefixed with the file name, e.g., `foo()` in `bar.daphne` could become `bar.foo()` in the importing script, to avoid name clashes.
- It would be okay to allow import statements only in the beginning of a DaphneDSL script.
- Parser errors should always indicate the actual file where the problem was encountered. So if `main.daphne` imports `bar.daphne` and there is a syntax error in `bar.daphne`, then the error message should report `bar.daphne`.

**Comment by akroviakov** [comment]
Assigned to @akroviakov .

**Comment by akroviakov** [comment]
_`foo()` in `bar.daphne` could become `bar.foo()` in the importing script, to avoid name clashes._
Identifier rule does not allow `.` in the name, could `_` be used as a delimeter for concatenating names with the prefix or must `.` be used?

_Importing a DaphneDSL script file should have roughly the same effect as pasting its contents literally._
The example with kmeans (`test/api/cli/algorithms/kmeans.daphne`) requires command line arguments, so I assume a function (i.e., `def km(r:si64,c:si64,f:si64,i:si64)`) was meant or am I wrong?

**Comment by akroviakov** [comment]
I'm trying to write unit test cases that use information from `UserConfig.json`.
An illustrative test case for one file (something similar to `ConfigTest.cpp`):
```c
const char* configFilePath = "test/api/cli/import/UserConfig.json";
DYNAMIC_SECTION(configFilePath) {
    compareDaphneToRef(
            "test/api/cli/import/import_success_4.txt",
            "test/api/cli/import/import_success_4.daphne", "--config", configFilePath
    );
}
```
It fails, so I added a check for user config existence:
```c
ConfigParser::fileExists(args["--config"])
```
And it showed `Parser error: could not open file '' for reading user config`.
I then took a look at what `LOG` in `Utils.h` says and it displays the following:
```
stdout: 
stderr: Parser error: could not open file '' for reading user config

status: 256
daphne --config test/api/cli/import/UserConfig.json test/api/cli/import/import_success_4.daphne
```
The member `args` in `DaphneDSLVisitor.cpp` has size 0, so I'm a bit confused.
Even if I simply take a test from `ConfigTest.cpp` checking for success and only change paths, it still displays this error.

Those scripts that do not _interact_ with UserConfig work fine regardless of whether `--config` was specified.
What would be the proper way to utilize UserConfig in unit test cases?

**Comment by akroviakov** [comment]
Please correct me if I'm wrong, but aren't `scriptArgsFinal` in `daphne.cpp` passed to `DaphneDSLParser`:
```cpp
DaphneDSLParser parser(scriptArgsFinal);
```
where they are stored as `args`:
```cpp
    DaphneDSLParser(std::unordered_map<std::string, std::string> args) : args(args) {
        //
    }
```
 which are then passed to the visitor in `DaphneDSLParser::parseStream()`:
```cpp
DaphneDSLVisitor visitor(module, builder, args);
```
and so on the console call like
```
./build.sh; build/bin/daphne test/api/cli/import/import_success_4.daphne --config=test/api/cli/import/UserConfig.json
```
we can access `--config` inside `DaphneDSLVisitor` through `args` and read it?

Also, I found what was the issue with:
```c
    compareDaphneToRef(
            "test/api/cli/import/import_success_4.txt",
            "test/api/cli/import/import_success_4.daphne", "--config", configFilePath
    );
```
It appears that in unit test cases  `"--config", configFilePath` are treated as separate arguments instead of one single argument, so we'd want to use `=` as daphne's `--help` suggests: 
```
--config test/api/cli/import/UserConfig.json
vs.
--config=test/api/cli/import/UserConfig.json
```
Do I misunderstand something about the current usage of `"--config", configFilePath` as arguments in `test/api/cli/config/ConfigTest.cpp`?

And a bit of a separate question, we need to read some paths in `DaphneDSLVisitor`, to get them we can:

1. Pass `DaphneUserConfig` from `daphne.cpp` all the way to `DaphneDSLVisitor` as an argument.
2. Rely on the already established `args` map, create `DaphneUserConfig` and read into it on imports.

The first option seems better than reading json on each import, but maybe you have some precautions against dragging `DaphneUserConfig` into the visitor?

**Comment by akroviakov** [comment]
Ok, I see my confusion with unit tests and config passing. 
In the above message I mentioned one call, and although this call is allowed, it is wrong:
```
./build.sh; build/bin/daphne test/api/cli/import/import_success_4.daphne --config=test/api/cli/import/UserConfig.json
```
since it is not conform to daphne's usage stated in `--help`:
```
daphne [options] script [arguments]
```
so a correct call would be:
```
./build.sh; build/bin/daphne --config=test/api/cli/import/UserConfig.json test/api/cli/import/import_success_4.daphne 
```
And just if someone would need the difference explained:
The first call gets `--config` into `scriptArgsFinal` in [`daphne.cpp`](https://github.com/daphne-eu/daphne/blob/main/src/api/cli/daphne.cpp), but `user_config` will be default.
The second call correctly fills `user_config` and leaves `scriptArgsFinal` empty.

What I was trying to do is to retrieve `--config` info from what is basically a copy of `scriptArgsFinal`. 
But since in unit testing (e.g., `compareDaphneToStr` in [Utils.h](https://github.com/daphne-eu/daphne/blob/main/test/api/cli/Utils.h)) we pass `args...` before `scriptFilePath`,  the `scriptArgsFinal` was empty and I was basically looking at an empty container. 

So in short, I haven't noticed that I violated daphne's usage stated in `--help` and got confused.

And yes, the question above whether to use `DaphneUserConfig` or `args` map is no longer relevant. `DaphneUserConfig` will be used to pass config to `DaphneDSLVisitor`.
